Feature/add crop screen

ajmaurya99 opened the pull request:
<!--
### Requirements

Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

- This PR attempts to add a crop screen after selecting an Avatar Image. Currently the suggested image dimensions: 200 by 200 pixels. 
- The crop screen is skipped if the Avatar size is  200 by 200 pixels.
- `Select Avatar` CTA is renamed to `Select Avatar and Crop`.
- The code is updated to ES6 standards.

### Screenshots

<img width="1736" alt="Screenshot 2021-09-29 at 4 49 07 PM" src="https://user-images.githubusercontent.com/9847340/135258802-e5f5b06a-b3cb-4698-bfa0-75e976e35af7.png">


### Alternate Designs

None

### Benefits

Now the users can crop their Avatar as they want it to appear on the website.

### Possible Drawbacks

Currently i havent used the `api.CroppedImageControl` and `api.SiteIconControl` but these can be used, in order to make the code more robust.

### Verification Process

1. Visit the user profile screen and scroll to the Avatar section at the bottom of the page.
2. Try updating/adding the Avatar image and you should be able to crop the Avatar Image as per your choice.

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Applicable Issues

https://github.com/10up/simple-local-avatars/issues/65

### Changelog Entry

<!-- Add sample CHANGELOG.md entry for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related. -->

Commit:
Initial cropping functionality added

Commit:
mediaUploader replaced with simple_local_avatar_frame

Commit:
labels corrected

Commit:
linting isues fixed and code updated to ES6

Antonio-Laguna commented on includes/class-simple-local-avatars.php:
```diff
@@ -272,14 +272,15 @@ public function admin_enqueue_scripts( $hook_suffix ) {
 
 		$script_name_append = ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) ? '.dev' : '';
-		wp_enqueue_script( 'simple-local-avatars', plugins_url( '', dirname( __FILE__ ) ) . '/simple-local-avatars' . $script_name_append . '.js', array( 'jquery' ), false, true );
+		wp_enqueue_script( 'simple-local-avatars', plugins_url( '', dirname( __FILE__ ) ) . '/simple-local-avatars.dev' . $script_name_append . '.js', array( 'jquery' ), false, true );
```
I think this should be reverted

Antonio-Laguna commented on includes/class-simple-local-avatars.php:
```diff
@@ -330,7 +331,7 @@ public function avatar_settings_field( $args ) {
 	 */
 	public function edit_user_profile( $profileuser ) {
-	?>
+		?>
```
Is this and below for linting? If it isn't I think it makes sense to revert

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -1,65 +1,164 @@
-			return;
+jQuery(document).ready(function ($) {
+	avatar_input = $('#simple-local-avatar');
+	avatar_preview = $('#simple-local-avatar-photo img');
+	current_avatar = avatar_preview.attr('src');
+	avatar_ratings = $('#simple-local-avatar-ratings');
+	avatar_container = $('#simple-local-avatar-photo');
+	$(document.getElementById('simple-local-avatar-media')).on('click', function (event) {
+		event.preventDefault();
 
-		if ( simple_local_avatar_frame ) {
-			simple_local_avatar_frame.open();
+		if (avatar_working) {
```
Missing spacing around parenthesis in all of the JS to follow WP guidelines

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -1,65 +1,164 @@
+		cropControl.mustBeCropped = function (flexW, flexH, dstW, dstH, imgW, imgH) {
+			// If the width and height are both flexible
+			// then the user does not need to crop the image.
+
```
Extra space for consistency

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -1,65 +1,164 @@
+					suggestedHeight: 200,
+				}),
+				new wp.media.controller.CustomizeImageCropper({
+					imgSelectOptions: myTheme_calculateImageSelectOptions,
```
I think here and below on the definition, we shouldn't namespace the function with `myTheme`

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -96,4 +195,115 @@ function avatar_lock( lock_or_unlock ) {
\ No newline at end of file
+}
+
+function myTheme_calculateImageSelectOptions(attachment, controller) {
```
Aside from the namespacing issue commented above, I think it makes sense to use `snake_case` for consistency with the rest of the functions defined in this file.

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -96,4 +195,115 @@ function avatar_lock( lock_or_unlock ) {
+				avatar_container.innerHTML = data;
+				jQuery(document.getElementById('simple-local-avatar-remove')).show();
+				avatar_ratings.disabled = false;
+				avatar_lock('unlock');
```
I think we should unlock it no matter what.

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -96,4 +195,115 @@ function avatar_lock( lock_or_unlock ) {
+		function (data) {
+			if (data != '') {
+				avatar_container.innerHTML = data;
+				jQuery(document.getElementById('simple-local-avatar-remove')).show();
```
Could be simplified

```suggestion
				jQuery('#simple-local-avatar-remove').show();
```

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -96,4 +195,115 @@ function avatar_lock( lock_or_unlock ) {
+		function (data) {
+			if (data != '') {
+				avatar_container.innerHTML = data;
+				jQuery(document.getElementById('simple-local-avatar-remove')).show();
```
Could be simplified

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -96,4 +195,115 @@ function avatar_lock( lock_or_unlock ) {
+				avatar_container.innerHTML = data;
+				jQuery(document.getElementById('simple-local-avatar-remove')).show();
+				avatar_ratings.disabled = false;
+				avatar_lock('unlock');
```
Same as above.

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -1,65 +1,164 @@
+	current_avatar = avatar_preview.attr('src');
+	avatar_ratings = $('#simple-local-avatar-ratings');
+	avatar_container = $('#simple-local-avatar-photo');
+	$(document.getElementById('simple-local-avatar-media')).on('click', function (event) {
```
Could be simplified

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -1,65 +1,164 @@
 	});
 
-	$( document.getElementById('simple-local-avatar-remove') ).on('click',function(event){
+	$(document.getElementById('simple-local-avatar-remove')).on('click', function (event) {
```
Could be simplified

Antonio-Laguna commented on simple-local-avatars.dev.js:
```diff
@@ -1,65 +1,164 @@
+			if (data != '') {
 				avatar_container.innerHTML = data;
-				$( document.getElementById('simple-local-avatar-remove') ).hide();
+				$(document.getElementById('simple-local-avatar-remove')).hide();
```
Could be simplified

Antonio-Laguna reviewed (commented):
Hey @ajmaurya99 !

This is amazing and I love all the comments you've left throughout the code. Everything is very well explained and the code is easy to follow.

I've added some comments as per your request to review this which has mostly formatting things.

Commit:
fix linting spacing issues

barneyjeffries committed:
revert change to dev script link

barneyjeffries committed:
- update cropping logic to auto skip if image matches
 - force square aspect ratio

Commit:
PR comments resolved for best practices

Commit:
composer file synced with the develop branch

helen commented on includes/class-simple-local-avatars.php:
```diff
@@ -329,8 +330,7 @@ public function avatar_settings_field( $args ) {
 	 */
-	public function edit_user_profile( $profileuser ) {
-	?>
+	public function edit_user_profile( $profileuser ) { ?>
```
```suggestion
	public function edit_user_profile( $profileuser ) { 
	?>
```

faisal-alvi replied:
@ajmaurya99 please add [this](https://github.com/10up/simple-local-avatars/pull/83#discussion_r719908468) suggestion by Helen.

helen commented on includes/class-simple-local-avatars.php:
```diff
@@ -415,7 +415,7 @@ class="button item-delete submitdelete deletion"
 		</table>
 	</div>
-	<?php
+		<?php
```
```suggestion
	<?php
```

helen commented on simple-local-avatars.dev.js:
```diff
@@ -1,99 +1,332 @@
+            return;
+         }
+
+         /* We need to setup a Crop control that contains a few parameters
```
Please remove comments that aren't relevant to the actual implementation - this one seems like help text from the example this came from.

helen commented on simple-local-avatars.dev.js:
```diff
@@ -1,99 +1,332 @@
+            return true;
+         };
+
+         /* NOTE: Need to set this up every time instead of reusing if already there
+         as the toolbar button does not get reset when doing the following:
+
+         simple_local_avatar_frame.setState('library');
+         simple_local_avatar_frame.open();
+         */
```
Remove unused code

helen commented on simple-local-avatars.dev.js:
```diff
@@ -1,99 +1,332 @@
-}
\ No newline at end of file
+   if ( undefined == avatar_spinner ) {
+      avatar_ratings = document.getElementById(
```
Reduce some of these formatting/whitespace changes.

helen reviewed (changes requested):
Works great! Could you possibly take a pass to reduce/remove the whitespace and other unrelated changes, as well as extraneous code comments that aren't needed for this context?

Commit:
Minified file updated along with best practices

Commit:
indentation added

ajmaurya99 commented:
@helen I have modified the PR based on your comments, Please have a look.

Commit:
wp scripts added

Commit:
removed package.json

helen commented on includes/class-simple-local-avatars.php:
```diff
@@ -272,14 +272,15 @@ public function admin_enqueue_scripts( $hook_suffix ) {
 
 		$script_name_append = ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) ? '.dev' : '';
-		wp_enqueue_script( 'simple-local-avatars', plugins_url( '', dirname( __FILE__ ) ) . '/simple-local-avatars' . $script_name_append . '.js', array( 'jquery' ), false, true );
+		wp_enqueue_script( 'simple-local-avatars', plugins_url( '', dirname( __FILE__ ) ) . '/build/index' . $script_name_append . '.js', array( 'jquery' ), false, true );
```
Would be ideal for this to still be named `simple-local-avatars.js` rather than `index.js`. It's nitpicky but since this could potentially have multiple JS files that don't all need to be loaded at once later I'd rather have this set up for specific file names from the start.

ajmaurya99 replied:
@faisal-alvi I have followed the architecture of the 'simple-podcasting' plugin and added the build file in the `dist` folder full path - `/dist/simple-local-avatars`

helen commented:
Another general thing here is that `.github/workflows/push-deploy.yml` will need some changes to account for the build, along with switching from `.gitattributes` to `.distignore` and deciding whether we want to include `src/**` in the .org deployment. See https://github.com/10up/simple-podcasting/ for examples.

faisal-alvi commented on includes/class-simple-local-avatars.php:
```diff
@@ -330,7 +331,7 @@ public function avatar_settings_field( $args ) {
 	 */
 	public function edit_user_profile( $profileuser ) {
-	?>
+		?>
```
Can you please also indent the code between these PHP tags which you indented?

faisal-alvi replied:
Can you please indent the content from Line 354 to 439?

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+		if (avatar_working) {
+			return;
+		}
+		/**
```
Please add a new line above the comment.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+		/**
+		 * Setup Crop control
+		 * The controls used by WordPress Admin are api.CroppedImageControl and api.SiteIconControl.
+		 */
```
Please remove the line below the comment.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+		 * as the toolbar button does not get reset when doing the following:
+		 * simple_local_avatar_frame.setState('library');
+		 * simple_local_avatar_frame.open();
+		 */
```
Please remove the line below the comment.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+		 * After the image has been cropped, apply the cropped image data to the setting
+		 *
+		 * @param {object} croppedImage Cropped attachment data.
+		 */
```
Please remove the line below the comment.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+
+	/**
+	 * Update the Local Avatar image
+	 */
```
Please remove the line below the comment.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+	});
+});
+
+function avatar_lock(lock_or_unlock) {
```
Can you please add short documentation above the function as you've already done very well for others?

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+	const x1 = (realWidth - xInit) / 2;
+	const y1 = (realHeight - yInit) / 2;
+
+	const imgSelectOptions = {
```
We can simply return from here by replacing the line with `return {` and removing the `return imgSelectOptions;` statement at the bottom.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+		});
+}
+
+function simple_local_avatar_set_image_from_attachment(attachment) {
```
Again, short documentation above the function.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+			_wpnonce: i10n_SimpleLocalAvatars.mediaNonce,
+		})
+		.done(function (data) {
+			if (data != '') {
```
Please replace with `!==`.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+			_wpnonce: i10n_SimpleLocalAvatars.mediaNonce,
+		})
+		.done(function (data) {
+			if (data != '') {
```
Please replace with `!==`.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+		avatar_form_button = jQuery(avatar_ratings).closest('form').find('input[type=submit]');
+	}
+
+	if (lock_or_unlock == 'unlock') {
```
Please replace with `===`.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+});
+
+function avatar_lock(lock_or_unlock) {
+	if (undefined == avatar_spinner) {
```
Please replace with `===`.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+			_wpnonce: i10n_SimpleLocalAvatars.deleteNonce,
+		})
+			.done(function (data) {
+				if (data != '') {
```
Please replace with `!==`.

faisal-alvi commented on src/index.js:
```diff
@@ -0,0 +1,336 @@
+		 */
+		cropControl.mustBeCropped = function (flexW, flexH, dstW, dstH, imgW, imgH) {
+			// Skip cropping if the image matches the crop dimension.
+			if (imgW === dstW && imgH === distH) {
```
Is the `distH` correct or should be replaced by `dstH`?

faisal-alvi reviewed (changes requested):
@ajmaurya99 Great work! the functionality is working as per the expectations.

1. Can you please work on the suggestions by Helen?
2. The `package.json` file was recently added via #94. Please resolve the conflicts.
3. Few more suggestions below, it's negligible but good to be done now than later, please check.

Commit:
PR comments resolved around code indentation

ajmaurya99 commented on package.json:
```diff
@@ -0,0 +1,29 @@
+    "name": "simple-local-avatars",
+    "version": "2.2.0",
+    "description": "Adds an avatar upload field to user profiles. Generates requested sizes on demand just like Gravatar!",
+    "scripts": {
```
Hey @faisal-alvi I have added WP Scripts to build the package.json and you have something like this,
could you help me here what codes need to be kept.

```
"scripts": {
    "build": "uglifyjs simple-local-avatars.dev.js -o simple-local-avatars.js -c -m"
  },
```

faisal-alvi replied:
@ajmaurya99 we can use `prebuild` to uglify JS file and keep `build` for `wp-scripts build`.

ajmaurya99 replied:
We dont need `prebuild` command as we have a build command to minify the `js` files.

Commit:
webpack config updated to include the dist folder

Commit:
build commands added in deploy pipeline and .gitattributes is replaced with .distignore

Commit:
sync with develop branch

Commit:
deleted old buid files

Commit:
deleted old buid files

ajmaurya99 commented:
> 



> Another general thing here is that `.github/workflows/push-deploy.yml` will need some changes to account for the build, along with switching from `.gitattributes` to `.distignore` and deciding whether we want to include `src/**` in the .org deployment. See https://github.com/10up/simple-podcasting/ for examples.

@faisal-alvi I have replaced the `.gitattributes` with `.distignore` and added the dist folder to .gitignore.

ajmaurya99 commented:
> @ajmaurya99 Great work! the functionality is working as per the expectations.
> 
> 1. Can you please work on the suggestions by Helen?
> 2. The `package.json` file was recently added via #94. Please resolve the conflicts.
> 3. Few more suggestions below, it's negligible but good to be done now than later, please check.

@faisal-alvi the current feature branch is synced with the develop branch.

faisal-alvi commented on includes/class-simple-local-avatars.php:
```diff
@@ -286,14 +286,15 @@ public function admin_enqueue_scripts( $hook_suffix ) {
 
 		$script_name_append = ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) ? '.dev' : '';
-		wp_enqueue_script( 'simple-local-avatars', plugins_url( '', dirname( __FILE__ ) ) . '/simple-local-avatars' . $script_name_append . '.js', array( 'jquery' ), false, true );
+		wp_enqueue_script( 'simple-local-avatars', plugins_url( '', dirname( __FILE__ ) ) . '/dist/simple-local-avatars' . $script_name_append . '.js', array( 'jquery' ), false, true );
```
The `$script_name_append` is no more required, should be removed otherwise it will throw a `404 error` (while trying to load `dist/simple-local-avatars.dev.js` file) when `SCRIPT_DEBUG` is `true`.

ajmaurya99 replied:
This is fixed @faisal-alvi

faisal-alvi commented on .distignore:
```diff
@@ -0,0 +1,37 @@
+/.git
```
If we are replacing the `.distignore` file with `.gitattributes`, then `.gitattributes` should be deleted. Correct?

ajmaurya99 replied:
I kind of missed that, I have removed the `.gitattributes` file.

faisal-alvi reviewed (changes requested):
@ajmaurya99 thanks for the changes, added a few small suggestions, can you please check?

Commit:
code indented and remove unused .distignore fie

faisal-alvi commented on includes/class-simple-local-avatars.php:
```diff
@@ -4,14 +4,15 @@
 	 */
 	public function __construct() {
-		$this->options        = (array) get_option( 'simple_local_avatars' );
+		 $this->options       = (array) get_option( 'simple_local_avatars' );
```
Please remove extra space from the beginning of the variable.

faisal-alvi reviewed (changes requested):
@ajmaurya99 a couple of very tiny changes, everything else is good! Thanks for the efforts!

Then we are good to go :rocket:.

Commit:
code indentation added

ajmaurya99 commented:
@faisal-alvi I have added the code indentation changes, Please have a look.

faisal-alvi reviewed (approved):
🚀
